src/jit/assertionprop.cpp

```diff
@@ -4605,18 +4605,39 @@ GenTree* Compiler::optPrepareTreeForReplacement(GenTree* oldTree, GenTree* newTr
 {
     // If we have side effects, extract them and append newTree to the list.
     GenTree* sideEffList = nullptr;
-    if (oldTree->gtFlags & GTF_PERSISTENT_SIDE_EFFECTS)
+    if ((oldTree->gtFlags & GTF_SIDE_EFFECT) != 0)
     {
-        gtExtractSideEffList(oldTree, &sideEffList, GTF_PERSISTENT_SIDE_EFFECTS_IN_CSE);
+        bool ignoreRoot = false;
+
+        if (oldTree == newTree)
+        {
+            // If the caller passed the same tree as both old and new then it means
+            // that it expects that the root of the tree has no side effects and it
+            // won't be extracted. Otherwise the resulting comma tree would be invalid,
+            // having both op1 and op2 point to the same tree.
+            //
+            // Do a sanity check to ensure persistent side effects aren't discarded and
+            // tell gtExtractSideEffList to ignore the root of the tree.
+            //
+            // Note that the check below is relaxed in order to allow discarding the
+            // root node even if it has exception side effects. It is assumed that
+            // the caller knows what it is doing (e.g. VN may be able to evaluate a
+            // DIV/MOD node to a constant and the node may still have GTF_EXCEPT set,
+            // even if it does not actually throw any exceptions).
+            assert(!gtNodeHasSideEffects(oldTree, GTF_PERSISTENT_SIDE_EFFECTS));
+            ignoreRoot = true;
```

Comment:
> It would be good to update GTF_EXCEPT flags once value numbering knows a tree is a constant but we currently don't have a convenient place to do that.

Yeah, I've tried something like that but things didn't went too well (e.g. `OperMayThrow` doesn't pay attention to VNs). Besides, I don't know if VN can be trusted with propagating exceptions, sometimes it packs constants with exceptions and sometimes it does not.

> Can you change the assert to:

I'll try, that looks like a reasonable addition.

Reply:
I made a typo in my suggestion. I meant replacing your assert with
```
assert(!gtNodeHasSideEffects(oldTree, GTF_SIDE_EFFECT) || vnStore->IsVNConstant(oldTree->gtVNPair.GetConservative())); 
```

Reply:
I added a separate assert for the GTF_EXCEPT | constant case, mostly because it was my addition of GTF_EXCEPT that triggered the need for this special handling. Considering the way the old code was working it seems that it is not possible for nodes with persistent side effects to become constants.